cpmpy/solvers/gcs.py

```diff
@@ -0,0 +1,413 @@
+        """
+        if is_num(cpm_expr):
+            return cpm_expr
+
+        # decision variables, check in varmap
+        if isinstance(cpm_expr, _NumVarImpl):  # _BoolVarImpl is subclass of _NumVarImpl
+            return self.solver_var(cpm_expr)
+
+        (obj_var, obj_cons) = get_or_make_var(cpm_expr)
+        self += obj_cons
+        return self.solver_var(obj_var)
+
+    def transform(self, cpm_expr):
+        """
+            Transform arbitrary CPMpy expressions to constraints the solver supports
+
+            Implemented through chaining multiple solver-independent **transformation functions** from
+            the `cpmpy/transformations/` directory.
+
+            See the 'Adding a new solver' docs on readthedocs for more information.
+
+        :param cpm_expr: CPMpy expression, or list thereof
+        :type cpm_expr: Expression or list of Expression
+
+        :return: list of Expression
+        """
+        # apply transformations, then post internally
+        # expressions have to be linearized to fit in MIP model. See /transformations/linearize
+        cpm_cons = toplevel_list(cpm_expr)
+        supported = {
+            "min", 
+            "max", 
+            "abs", 
+            "alldifferent", 
+            "element", 
+            'table', 
+            'negative_table', 
+            'count', 
+            'nvalue',
+            'inverse', 
+            'circuit', 
+            'xor'}
+        cpm_cons = decompose_in_tree(cpm_cons, supported)
+        cpm_cons = flatten_constraint(cpm_cons)  # flat normal form
+
+        # NB: GCS supports full reification for linear equality and linear inequaltiy constraints
+        # but no reification for linear not equals and not half reification for linear equality. 
+        # Maybe a future transformation (or future work on the GCS solver).
+        cpm_cons = reify_rewrite(cpm_cons, supported=frozenset(['==']))
+        cpm_cons = only_numexpr_equality(cpm_cons, supported=frozenset(["sum", "wsum"]))  # supports >, <, !=
+
+        # NB: GCS supports a small number of simple expressions as the reifying term
+        # e.g. (x > 3) -> constraint could in principle be supported in future. 
+        cpm_cons = only_bv_reifies(cpm_cons)
+        return cpm_cons
+
+    def __add__(self, cpm_cons):
+        """
+        Post a (list of) CPMpy constraints(=expressions) to the solver
+        Note that we don't store the constraints in a cpm_model,
+        we first transform the constraints into primitive constraints,
+        then post those primitive constraints directly to the native solver
+        :param cpm_con CPMpy constraint, or list thereof
+        :type cpm_con (list of) Expression(s)
+        """
+        # add new user vars to the set
+                # add new user vars to the set
+        get_variables(cpm_cons, collect=self.user_vars)
+
+        for con in self.transform(cpm_cons):
+            self._post_constraint(con)
```

Comment:
for tiny efficiency speedups, we recommend putting the body of _post_constraint here instead of doing a function call for every element in the list; that is what we now do in the other solvers too.